lms/djangoapps/certificates/queue.py

```diff
@@ -306,6 +307,14 @@ def add_cert(self, student, course_id, course=None, forced_grade=None, template_
                 #   Despite blowing up the xml parser, bad values here are fine
                 grade_contents = None
 
+            # If this user's enrollment is not eligible to receive a
+            # certificate, mark it as such for reporting and
+            # analytics.
+            if not is_whitelisted and not CourseMode.is_eligible_for_certificate(enrollment_mode):
+                cert.eligible_for_certificate = False
+                cert.status = CertificateStatuses.auditing
+                cert.save()
+
             if is_whitelisted or grade_contents is not None:
```

Comment:
Like here -- if we can somehow break up conditionals like this the code would (hopefully) become simpler and likely easier to test.